packages/components/src/components/DetailsReferences.vue

```diff
@@ -0,0 +1,154 @@
+<template>
+  <div class="details-references">
+    <div class="details-references__header">
+      <h4 class="details-references__header-title">Find references</h4>
+      <v-list-item-param
+        :name="param.name"
+        :value="param.value"
+        :className="param.class"
+      />
+    </div>
+    <v-list title="Matches">
+      <v-list-item
+        class="details-references__item"
+        v-for="item in matchedEvents"
+        :key="item.event.id"
+        @click.native="focusEvent(item.event)"
+        @dblclick.native="selectEvent(item.event)"
+      >
+        {{ item.event.toString() }}
+        <div class="details-references__badges">
+          <span
+            v-if="item.match.includes('value')"
+            class="details-references__badge"
+            >Value</span
+          >
+          <span
+            v-if="item.match.includes('reference')"
+            class="details-references__badge"
+            >Reference</span
+          >
+        </div>
+      </v-list-item>
+    </v-list>
+  </div>
+</template>
+
+<script>
+import { AppMap } from '@appland/models';
+import VList from '@/components/common/List.vue';
+import VListItem from '@/components/common/ListItem.vue';
+import VListItemParam from '@/components/common/ListItemParam.vue';
+import { SELECT_OBJECT, SET_FOCUSED_EVENT } from '@/store/vsCode';
+
+export default {
+  name: 'v-details-references',
+  components: {
+    VList,
+    VListItem,
+    VListItemParam,
+  },
+  props: {
+    param: {
+      type: Object,
+      required: true,
+    },
+    appMap: AppMap,
+  },
+  computed: {
+    matchedEvents() {
+      const events = {};
+      const regex = new RegExp(this.param.value, 'i');
+
+      this.appMap.events.forEach((event) => {
+        if (!event.isCall()) {
+          return;
+        }
+        // find by value
+        if (
+          (event.parameters &&
+            event.parameters.some((param) => regex.test(param.value))) ||
+          (event.message &&
+            event.message.some((msg) => regex.test(msg.value))) ||
+          (event.sqlQuery && regex.test(event.sqlQuery)) ||
+          (event.returnValue && regex.test(event.returnValue.value))
+        ) {
+          events[event.id] = { event, match: ['value'] };
+        }
+        // find by reference
+        if (
+          (event.parameters &&
+            event.parameters.some(
+              (param) => this.param.object_id === param.object_id
+            )) ||
+          (event.message &&
+            event.message.some(
+              (msg) => this.param.object_id === msg.object_id
+            )) ||
+          (event.returnValue &&
+            this.param.object_id === event.returnValue.object_id)
+        ) {
+          if (events[event.id]) {
+            events[event.id].match.push('reference');
+          } else {
+            events[event.id] = { event, match: ['reference'] };
+          }
+        }
+      });
```

Comment:
This could use some sort of abstraction to clean it up. Reducing the amount of branches and logical ORs would really help. I've added some pseudo-code below, just kind of spit-balling, but maybe it will help provide some direction.

```ts
type EventValueType = 'sql' | 'param' | 'return_value';

interface EventValue {
  eventId: number;
  objectId?: number;
  value: string,
  valueType: EventValueType
}

interface EventValueReference {
  eventId: number,
  valueType: EventValueType,
  byReference?: boolean,
  byValue?: boolean,
  name?: string
}

function findReference(obj: EventValue, searchTerm: string, objectId?: number) -> EventValueReference | undefined {
  let byReference = objectId !== undefined && objectId === obj.objectId;
  let byValue = obj.value.includes(searchTerm);
  if (byReference || byValue) {
    return { /* EventValueReference properties */ };
  }   
}

const createValueArray = (array, valueType) => (array || []).map((obj) => ({eventId: event.id, objectId: obj.object_id, value: obj.value, valueType }));

// Normalize the values we'll be searching on into a single array
const values = [
  ...createValueArray(event.parameters, 'param'),
  ...createValueArray(event.message, 'param'),
  (event.returnValue && {eventId: event.id, objectId: event.returnValue.objectId, value: event.returnValue.value, valueType: 'return_value'}),
  (event.sql && {eventId: event.id, value: event.sqlQuery, valueType: 'sql'}),
];

// Perform the reference search and return results
return values.map((val) => findReference(val, searchTerm, objectId)).filter(Boolean);
```